estimator/sis_lattice.py

```diff
@@ -0,0 +1,331 @@
+
+        r = simulator(d=d_, n=d_ - params.n, q=params.q, beta=beta, xi=1, tau=False)
+
+        # Cost the sampling of short vectors.
+        rho, cost_red, N, sieve_dim = red_cost_model.short_vectors(beta, d_)
+        bkz_cost = costf(red_cost_model, beta, d_)
+
+        if RR(sqrt(d)) * params.length_bound <= params.q:  # Non-dilithium style analysis
+            # Calculate expected vector length using approximation factor on the shortest vector from BKZ
+            vector_length = rho * sqrt(r[0])
+            # Find probability that all coordinates meet norm bound
+            sigma = vector_length / sqrt(d_)
+            log_trial_prob = RR(d_*log(1 - 2*gaussian_cdf(0, sigma, -params.length_bound), 2))
+
+        else:  # Dilithium style analysis
+            # Find first non-q-vector in r
+            if abs(sqrt(r[0]) - params.q) < 1e-8:  # q-vectors exist
+                idx_start = next(i for i, r_ in enumerate(r) if r_ < r[0])
+
+            else:
+                idx_start = 0
+
+            if abs(r[-1] - 1) < 1e-8:  # 1-vectors exist
+                # Find first 1 length graham-schmidt vector in r (Zone III)
+                idx_end = next((i - 1 for i, r_ in enumerate(r) if sqrt(r_) <= 1 + 1e-8), d_ - 1)
+
+            else:
+                idx_end = d_ - 1
+
+            vector_length = sqrt(r[idx_start])
+            gaussian_coords = max(idx_end - idx_start + 1, sieve_dim)
+            sigma = vector_length / sqrt(gaussian_coords)
+
+            log_trial_prob = RR(log(1 - 2*gaussian_cdf(0, sigma, -params.length_bound), 2)*(gaussian_coords))
+            log_trial_prob += RR(log((2*params.length_bound + 1)/params.q, 2)*(idx_start))
+
+        probability = 2**min(0, log_trial_prob + RR(log(N, 2)))  # expected number of solutions (max 1)
+        ret = Cost()
+        ret["rop"] = cost_red
+        ret["red"] = bkz_cost["rop"]
+        ret["sieve"] = max(cost_red - bkz_cost["rop"], 1e-100)  # Ensuring non-zero cost here
+        ret["beta"] = beta
+        ret["eta"] = sieve_dim
+        ret["zeta"] = zeta
+        ret["d"] = d_
+        ret["prob"] = probability
+
+        ret.register_impermanent(
+            rop=True,
+            red=True,
+            sieve=True,
+            eta=False,
+            zeta=False,
+            prob=False,
+        )
+        # 4. Repeat whole experiment ~1/prob times
+        if probability and not RR(probability).is_NaN():
+            ret = ret.repeat(
+                prob_amplify(success_probability, probability),
+            )
+        else:
+            return Cost(rop=oo)
+
+        return ret
+
+    @classmethod
+    def cost_zeta(
+        cls,
+        zeta: int,
+        params: SISParameters,
+        ignore_qary: bool = False,
+        red_shape_model=red_simulator_default,
+        red_cost_model=red_cost_model_default,
+        d=None,
+        log_level=5,
+        **kwds,
+    ):
+        """
+        This function optimizes costs for a fixed guessing dimension ζ.
```

Comment:
Guessing dimension might be a bit misleading?

Reply:
Changed to make a bit more descriptive.